### ISSUE TITLE
Question about delta position

### ISSUE THREAD

**Issue by kasia-te** [Kuanhao-Chao/OpenSpliceAI#9]
Hi,
I'm trying the OpenSplice AI variant.
My command is:
```
openspliceai variant \
        --input-vcf input.vcf.gz \
        --ref-genome GRCh38.fa \
        --annotation grch38.txt \
        --model models/ \
        --model-type pytorch \
        --flanking-size 10000 \
        --mask 0 \
        --output-vcf ./annotated.vcf.gz
``` 
 I have a question about the output. Even though I am using the default distance of 50, the delta positions shown in the output are much longer, for example:
```
chr8    140300616       chr8_140300616_T_G      T       G       48      .       SpliceAI=G|TRAPPC9|0.06|0.95|0.00|0.79|4968|4998|1875|4853 
```
Why is that?

Also, how can I prepare a custom annotation file?
Thank you,
Kasia

**Comment by conbward** [comment]
I have also had this same issue with the output of the variant command. 
```
openspliceai variant \
 -R /hg38.fa \
 -A grch38.txt  \
 -m OSAI-MANE/10000nt/ \
 -f 400 \
 -t pytorch \
 -I test.vcf \
 -O output/test.vcf
```

Although I am using the default of 50, most of the delta positions are in the 200s, and there are only positive delta positions (no negative). I've also noticed that the scores for SNVs at the canonical splice acceptor/donor motifs are lower than expected (as compared to the original SpliceAI scores, for example the original SpliceAI gives these SNVs scores of .99 for acceptor loss, but OSAI scores are closer to ~.45). 

For example this is a canonical splice acceptor. You can note both the lower scores and also the mislabeled positions (you would expect the acceptor loss delta positions to be 2 and 1, not 202 and 201. 

```
2	47466656	.	A	C	.	.	SpliceAI=C|MSH2|0.23|0.46|0.01|0.00|222|202|201|270
2	47466656	.	A	G	.	.	SpliceAI=G|MSH2|0.53|0.48|0.01|0.01|201|202|201|202
2	47466656	.	A	T	.	.	SpliceAI=T|MSH2|0.23|0.47|0.00|0.01|222|202|208|270
2	47466657	.	G	A	.	.	SpliceAI=A|MSH2|0.36|0.48|0.00|0.01|202|201|202|201
2	47466657	.	G	C	.	.	SpliceAI=C|MSH2|0.19|0.41|0.00|0.00|221|201|207|155
2	47466657	.	G	T	.	.	SpliceAI=T|MSH2|0.22|0.46|0.00|0.01|221|201|192|269
```

As a note, this same delta position error can be seen in the example output.vcf provided in the repository (examples/variant/output.vcf).

Based on the comment above and my output, I am wondering if the error has something to do with the flanking size provided as my delta positions differ from expected by ~200bp and the example above differ by ~5000, both of which are half of the flanking size argument used to run the command.  

Thanks!
Connor

**Comment by kasia-te** [comment]
It seems that this position is measured from the start of the flanking region, rather than from the variant position.
In my example, it should be -2 (https://spliceailookup.broadinstitute.org/#variant=chr8-140300616-T-G&hg=38&bc=basic&distance=10000&mask=0&ra=0).
Therefore, it is probably sufficient to subtract half of the flanking length from the delta values to obtain the correct answer.

Anyway, I checked the first variant of your example using the `-f 10000` argument, and the DS_AL is 0.99. I'm using  OSAIMANE-10000nt models given [here](https://ccb.jhu.edu/openspliceai/content/pretrained_models/GRCh38_MANE.html#human-mane-spliceai)  

```
chr2    47466656        .      A       C       .     .       SpliceAI=C|MSH2|0.48|0.99|0.00|0.34|5022|5002|4842|5152
```

Thanks!
Kasia

**Comment by conbward** [comment]
I don't think subtracting half the flanking length from the delta values is sufficient in all cases, as it still gives you positions outside the 50bp window that it should be searching for. Using the line above as an example, the donor gain/loss position are both beyond the 50bp window. 

```
2	47466656	.	A	C	.	.	SpliceAI=C|MSH2|0.48|0.99|0.00|0.34|22|2|-158|152
```

With the line from the original post you see something similar. None of these distances should be greater than 50bp from the variant. 
```
chr8    140300616       chr8_140300616_T_G      T       G       48      .       SpliceAI=G|TRAPPC9|0.06|0.95|0.00|0.79|-32|-2|-3125|147
```

**Comment by kasia-te** [comment]
You are right.
I suppose we'll have to wait for the developers' response.

Best,
Kasia

wt., 27 maj 2025 o 18:15 conbward ***@***.***> napisał(a):

**Comment by conbward** [comment]
As a follow-up I believe the issue here may be that the flanking sequence is never removed (the comment in line 493 of variant/utils.py indicates it is but I don't think that's happening, and if so that would only be in the pytorch based model if I'm reading it rigth). One work around we found was to add the following code in the get_delta_scores funtion right before idx_pa and the others are retrieved (currently line 525)

''' 
 
            middle_index = y.shape[1] // 2
            start_index = middle_index - dist_var
            end_index = middle_index + dist_var + 1  # +1 because the end index is exclusive in slicing

            # Slice y to contain only the window 
            y = y[:, start_index:end_index, :]
'''

Thanks!

**Comment by Kuanhao-Chao** [comment]
Hi @kasia-te and @conbward ,
Thank you both for the detailed reports, this is extremely helpful.

You are correct: the issue comes from the flanking window not being cropped before extracting the delta positions. As a result, the indices were being computed relative to the full flanking region rather than the variant-centered 50 bp window, which explains the large offsets you observed.

I’ve now fixed this cropping bug in the variant module so that the model output is properly trimmed around the variant before delta coordinates are calculated. This resolves the incorrect delta positions seen in both the example output and your runs. The fix will be included in the upcoming release.

Thanks again for catching this and for your patience!
Best,
Kuan-Hao

**Comment by Kuanhao-Chao** [comment]
@kasia-te 

I’ve created a script to help you convert your GFF file into the annotation format required by OpenSpliceAI.  
Please see: https://github.com/Kuanhao-Chao/OpenSpliceAI/tree/main/data

Best,  
Kuan-Hao
